File: lib/python/opentoken/src/test/opentoken/attributes/person/us_postal_code_attribute_test.py
"""
Tests for USPostalCodeAttribute.
"""

import pickle
import pytest
from concurrent.futures import ThreadPoolExecutor, as_completed
from opentoken.attributes.person.us_postal_code_attribute import USPostalCodeAttribute


class TestUSPostalCodeAttribute:
    """Test cases for USPostalCodeAttribute."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.us_postal_code_attribute = USPostalCodeAttribute()

    def test_get_name_should_return_us_postal_code(self):
        """Test that get_name returns 'USPostalCode'."""
        assert self.us_postal_code_attribute.get_name() == "USPostalCode"

    def test_get_aliases_should_return_us_zip_code_aliases(self):
        """Test that get_aliases returns USPostalCode and USZipCode."""
        expected_aliases = ["USPostalCode", "USZipCode"]
        assert self.us_postal_code_attribute.get_aliases() == expected_aliases

    def test_normalize_should_return_first_5_digits(self):
        """Test normalization returns first 5 digits for US ZIP codes."""
        assert self.us_postal_code_attribute.normalize("10001-6789") == "10001"
        assert self.us_postal_code_attribute.normalize("10001") == "10001"
        assert self.us_postal_code_attribute.normalize("100016789") == "10001"
        assert self.us_postal_code_attribute.normalize("95123-6789") == "95123"
        assert self.us_postal_code_attribute.normalize("951236789") == "95123"
        assert self.us_postal_code_attribute.normalize("65201-6789") == "65201"
        assert self.us_postal_code_attribute.normalize("652016789") == "65201"

    def test_normalize_should_handle_whitespace(self):
        """Test normalization handles whitespace for US ZIP codes."""
        assert self.us_postal_code_attribute.normalize("10001") == "10001"
        assert self.us_postal_code_attribute.normalize(" 10001") == "10001"
        assert self.us_postal_code_attribute.normalize("10001 ") == "10001"
        assert self.us_postal_code_attribute.normalize(" 10001 ") == "10001"
        assert self.us_postal_code_attribute.normalize("1 0 0 0 1") == "10001"
        assert self.us_postal_code_attribute.normalize("10\t001") == "10001"
        assert self.us_postal_code_attribute.normalize("10\n001") == "10001"
        assert self.us_postal_code_attribute.normalize("10\r\n001") == "10001"
        assert self.us_postal_code_attribute.normalize("  10   001  ") == "10001"

    def test_validate_should_return_true_for_valid_us_zip_codes(self):
        """Test validation returns true for valid US ZIP codes."""
        assert self.us_postal_code_attribute.validate("95123 ") is True
        assert self.us_postal_code_attribute.validate(" 95123") is True
        assert self.us_postal_code_attribute.validate("95123") is True
        assert self.us_postal_code_attribute.validate("95123-6789") is True
        assert self.us_postal_code_attribute.validate("951236789") is True
        assert self.us_postal_code_attribute.validate("65201-6789") is True
        assert self.us_postal_code_attribute.validate("652016789") is True
        assert self.us_postal_code_attribute.validate("10001") is True
        assert self.us_postal_code_attribute.validate("100016789") is True
        assert self.us_postal_code_attribute.validate("90210-1234") is True
        assert self.us_postal_code_attribute.validate("902101234") is True

    def test_validate_should_return_false_for_null_and_empty(self):
        """Test validation returns false for null and empty values."""
        assert self.us_postal_code_attribute.validate(None) is False
        assert self.us_postal_code_attribute.validate("") is False

    def test_validate_should_return_false_for_invalid_formats(self):
        """Test validation returns false for invalid formats."""
        assert self.us_postal_code_attribute.validate("1234") is False
        assert self.us_postal_code_attribute.validate("123456") is False
        assert self.us_postal_code_attribute.validate("1234-5678") is False
        assert self.us_postal_code_attribute.validate("abcde") is False
        assert self.us_postal_code_attribute.validate("K1A 0A6") is False

    def test_validate_should_return_false_for_placeholder_zip_codes(self):
        """Test validation returns false for placeholder ZIP codes."""
        invalid_zip_codes = [
            "12345", "54321", "01234", "98765", "00000", "11111", "22222",
            "33333", "55555", "66666", "77777", "88888", "99999"
        ]
        for zip_code in invalid_zip_codes:
            assert self.us_postal_code_attribute.validate(zip_code) is False

    def test_normalize_thread_safety(self):
        """Test thread safety of normalize method."""
        thread_count = 100
        test_postal_code = "10001-6789"
        expected_result = "10001"
        results = []

        def normalize_postal_code():
            return self.us_postal_code_attribute.normalize(test_postal_code)

        with ThreadPoolExecutor(max_workers=thread_count) as executor:
            futures = [executor.submit(normalize_postal_code) for _ in range(thread_count)]
            for future in as_completed(futures):
                results.append(future.result())

        assert len(results) == thread_count
        assert all(result == expected_result for result in results)

    def test_normalize_should_handle_edge_cases(self):
        """Test normalization handles edge cases."""
        assert self.us_postal_code_attribute.normalize("1234 ") == "1234"
        assert self.us_postal_code_attribute.normalize("123") == "123"
        assert self.us_postal_code_attribute.normalize("12") == "12"
        assert self.us_postal_code_attribute.normalize("1") == "1"
        assert self.us_postal_code_attribute.normalize(None) is None
        assert self.us_postal_code_attribute.normalize("") == ""
        assert self.us_postal_code_attribute.normalize("10001") == "10001"
        assert self.us_postal_code_attribute.normalize(" 10001") == "10001"
        assert self.us_postal_code_attribute.normalize("10001-1234") == "10001"
        assert self.us_postal_code_attribute.normalize("100011234") == "10001"
        assert self.us_postal_code_attribute.normalize("K1A 0A7") == "K1A 0A7"
        assert self.us_postal_code_attribute.normalize("k1a0a7") == "k1a0a7"

    def test_serialization(self):
        """Test serialization and deserialization of USPostalCodeAttribute."""
        serialized_data = pickle.dumps(self.us_postal_code_attribute)
        deserialized_attribute = pickle.loads(serialized_data)

        test_values = [
            "10001", "10001-1234", "90210-5678", "30301", "60601-2345"
        ]

        for value in test_values:
            assert self.us_postal_code_attribute.get_name() == deserialized_attribute.get_name()
            assert self.us_postal_code_attribute.get_aliases() == deserialized_attribute.get_aliases()
            assert self.us_postal_code_attribute.normalize(value) == deserialized_attribute.normalize(value)
            assert self.us_postal_code_attribute.validate(value) == deserialized_attribute.validate(value)

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/attributes/person/USPostalCodeAttributeTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.person;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.TimeUnit;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class USPostalCodeAttributeTest {
    private USPostalCodeAttribute usPostalCodeAttribute;

    @BeforeEach
    void setUp() {
        usPostalCodeAttribute = new USPostalCodeAttribute();
    }

    @Test
    void getName_ShouldReturnUSPostalCode() {
        assertEquals("USPostalCode", usPostalCodeAttribute.getName());
    }

    @Test
    void getAliases_ShouldReturnUSZipCodeAliases() {
        String[] expectedAliases = { "USPostalCode", "USZipCode" };
        assertArrayEquals(expectedAliases, usPostalCodeAttribute.getAliases());
    }

    @Test
    void normalize_ShouldReturnFirst5Digits() {
        assertEquals("10001", usPostalCodeAttribute.normalize("10001-6789"));
        assertEquals("10001", usPostalCodeAttribute.normalize("10001"));
        assertEquals("10001", usPostalCodeAttribute.normalize("100016789"));
        assertEquals("95123", usPostalCodeAttribute.normalize("95123-6789"));
        assertEquals("95123", usPostalCodeAttribute.normalize("951236789"));
        assertEquals("65201", usPostalCodeAttribute.normalize("65201-6789"));
        assertEquals("65201", usPostalCodeAttribute.normalize("652016789"));
    }

    @Test
    void normalize_ShouldHandleWhitespace() {
        // Test different types of whitespace for US ZIP codes
        assertEquals("10001", usPostalCodeAttribute.normalize("10001"), "No whitespace");
        assertEquals("10001", usPostalCodeAttribute.normalize(" 10001"), "Leading space");
        assertEquals("10001", usPostalCodeAttribute.normalize("10001 "), "Trailing space");
        assertEquals("10001", usPostalCodeAttribute.normalize(" 10001 "), "Leading and trailing spaces");
        assertEquals("10001", usPostalCodeAttribute.normalize("1 0 0 0 1"), "Spaces between digits");
        assertEquals("10001", usPostalCodeAttribute.normalize("10\t001"), "Tab character");
        assertEquals("10001", usPostalCodeAttribute.normalize("10\n001"), "Newline character");
        assertEquals("10001", usPostalCodeAttribute.normalize("10\r\n001"), "Carriage return and newline");
        assertEquals("10001", usPostalCodeAttribute.normalize("  10   001  "), "Multiple spaces");
    }

    @Test
    void validate_ShouldReturnTrueForValidUSZipCodes() {
        assertTrue(usPostalCodeAttribute.validate("95123 "));
        assertTrue(usPostalCodeAttribute.validate(" 95123"));
        assertTrue(usPostalCodeAttribute.validate("95123"));
        assertTrue(usPostalCodeAttribute.validate("95123-6789"));
        assertTrue(usPostalCodeAttribute.validate("951236789"));
        assertTrue(usPostalCodeAttribute.validate("65201-6789"));
        assertTrue(usPostalCodeAttribute.validate("652016789"));
        assertTrue(usPostalCodeAttribute.validate("10001"));
        assertTrue(usPostalCodeAttribute.validate("100016789"));
        assertTrue(usPostalCodeAttribute.validate("90210-1234"));
        assertTrue(usPostalCodeAttribute.validate("902101234"));
    }

    @Test
    void validate_ShouldReturnFalseForNullAndEmpty() {
        assertFalse(usPostalCodeAttribute.validate(null), "Null value should not be allowed");
        assertFalse(usPostalCodeAttribute.validate(""), "Empty value should not be allowed");
    }

    @Test
    void validate_ShouldReturnFalseForInvalidFormats() {
        assertFalse(usPostalCodeAttribute.validate("1234"), "Short postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("123456"), "Long postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("1234-5678"), "Invalid format should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("abcde"), "Non-numeric should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("K1A 0A6"), "Canadian postal code should not validate");
    }

    @Test
    void validate_ShouldReturnFalseForPlaceholderZipCodes() {
        assertFalse(usPostalCodeAttribute.validate("12345"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("54321"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("01234"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("98765"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("00000"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("11111"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("22222"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("33333"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("55555"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("66666"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("77777"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("88888"), "Invalid postal code should not be allowed");
        assertFalse(usPostalCodeAttribute.validate("99999"), "Invalid postal code should not be allowed");
    }

    @Test
    void normalize_ThreadSafety() throws InterruptedException {
        final int threadCount = 100;
        final CountDownLatch startLatch = new CountDownLatch(1);
        final CountDownLatch finishLatch = new CountDownLatch(threadCount);
        final CyclicBarrier barrier = new CyclicBarrier(threadCount);
        final List<String> results = Collections.synchronizedList(new ArrayList<>());
        final String testPostalCode = "10001-6789";

        for (int i = 0; i < threadCount; i++) {
            Thread thread = new Thread(() -> {
                try {
                    startLatch.await();
                    barrier.await();

                    String result = usPostalCodeAttribute.normalize(testPostalCode);
                    results.add(result);

                    finishLatch.countDown();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
            thread.start();
        }

        startLatch.countDown();
        finishLatch.await(15, TimeUnit.SECONDS);

        assertEquals(threadCount, results.size());
        for (String result : results) {
            assertEquals("10001", result);
        }
    }

    @Test
    void normalize_ShouldHandleEdgeCases() {
        // Test short postal codes (less than 5 characters) - should return trimmed original
        assertEquals("1234", usPostalCodeAttribute.normalize("1234 "));
        assertEquals("123", usPostalCodeAttribute.normalize("123"));
        assertEquals("12", usPostalCodeAttribute.normalize("12"));
        assertEquals("1", usPostalCodeAttribute.normalize("1"));

        // Test null and empty values
        assertNull(usPostalCodeAttribute.normalize(null));
        assertEquals("", usPostalCodeAttribute.normalize(""));

        // Test exactly 5 characters and longer ZIP codes
        assertEquals("10001", usPostalCodeAttribute.normalize("10001"));
        assertEquals("10001", usPostalCodeAttribute.normalize(" 10001"));
        assertEquals("10001", usPostalCodeAttribute.normalize("10001-1234"));
        assertEquals("10001", usPostalCodeAttribute.normalize("100011234"));

        // Test non-US formats - should return trimmed original
        assertEquals("K1A 0A7", usPostalCodeAttribute.normalize("K1A 0A7"));
        assertEquals("k1a0a7", usPostalCodeAttribute.normalize("k1a0a7"));
    }

    @Test
    void testSerialization() throws Exception {
        // Serialize the attribute
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(baos);
        oos.writeObject(usPostalCodeAttribute);
        oos.close();

        // Deserialize the attribute
        ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
        ObjectInputStream ois = new ObjectInputStream(bais);
        USPostalCodeAttribute deserializedAttribute = (USPostalCodeAttribute) ois.readObject();
        ois.close();

        // Test various US ZIP code values with both original and deserialized attributes
        String[] testValues = {
                "10001",
                "10001-1234",
                "90210-5678",
                "30301",
                "60601-2345"
        };

        for (String value : testValues) {
            assertEquals(
                    usPostalCodeAttribute.getName(),
                    deserializedAttribute.getName(),
                    "Attribute names should match");

            assertArrayEquals(
                    usPostalCodeAttribute.getAliases(),
                    deserializedAttribute.getAliases(),
                    "Attribute aliases should match");

            assertEquals(
                    usPostalCodeAttribute.normalize(value),
                    deserializedAttribute.normalize(value),
                    "Normalization should be identical for value: " + value);

            assertEquals(
                    usPostalCodeAttribute.validate(value),
                    deserializedAttribute.validate(value),
                    "Validation should be identical for value: " + value);
        }
    }
}